### ISSUE TITLE
Support translations and setup weblate to accept translations

### ISSUE THREAD

**Issue by robe2** [postgis/postgis-workshops#30]
I thought postgis-workshops being a user facing piece should really have translations.
I'm working thru this https://www.sphinx-doc.org/en/master/usage/advanced/intl.html to see if I can fit it in.

The plan is to allow users who prefer git to make changes directly in git and other users to use weblate which I have installed here https://weblate.osgeo.org

For authentication with weblate, I will start with osgeo LDAP first, but then plan to open it up to github/gitlab users once I've done some more testing to make sure I can support multiple auth channels.

Eventually I'd like to switch postgis translations from transifex to weblate, but don't want to mess with that until I've thoroughly tested it out with a virgin project like postgis-workshops.

**Comment by pramsey** [comment]
Very ambitious! I hope it works out!

**Comment by robe2** [comment]
> Very ambitious! I hope it works out!

Yap ambitious lets see how my energy holds up. Maybe I can get @strk interested again.

I got @dkastl interested in using weblate and he has some experience with it so that will energize me a bit.  He's already setup a new weblate project - https://weblate.osgeo.org/projects/gtt-project/

on the workshop side -- I'm still fighting with making the pos fall in separate folders.  Right now when I build it just overwrites the english.  Other issue I have is the po updates coming back from weblate seem to be a little different than what python build script is expecting, so giving me some random compile errors.  Seems to be with the way it breaks the lines.

**Comment by dkastl** [comment]
Well, my experience with Weblate is that I couldn't manage to get it working to push to Github. But one reason was probably, that I didn't have enough permissions on the Github repository. So I will try now again with the repository I have admin access for.

The hosted Weblate for Libre Projects has some more version systems listed, for example Github: 
![Screenshot from 2021-11-29 15-44-01](https://user-images.githubusercontent.com/227762/143820783-2e821282-b924-42f6-bb82-eff62272f396.png)
Not sure if this is a configuration option. Eventually it simplifies the setup with Github.

**Comment by dkastl** [comment]
@robe2 , to be able to push to Github you mentioned the SSH key: https://weblate.osgeo.org/keys/

Now Weblate says, that creating a Github user with this SSH Key is a way to grant push access: https://docs.weblate.org/en/latest/vcs.html#verifying-ssh-host-keys

However, when I try to do this, Github returns `Key is already in use`. This means, there can be only one Github user, that projects could add to their projects. Do you know, which Github user this is? Hosted Weblate seems to do something similar: https://docs.weblate.org/en/latest/vcs.html#accessing-repositories-from-hosted-weblate

**Comment by robe2** [comment]
@dkastl  I didn't need to create a github user to push.  What I did is register the Weblate ssh key as a deploy key in the project.  I was testing with my local fork but will add to this one now.

**Comment by robe2** [comment]
> Now Weblate says, that creating a Github user with this SSH Key is a way to grant push access: https://docs.weblate.org/en/latest/vcs.html#verifying-ssh-host-keys
> 
> However, when I try to do this, Github returns `Key is already in use`. This means, there can be only one Github user, that projects could add to their projects. Do you know, which Github user this is? Hosted Weblate seems to do something similar: https://docs.weblate.org/en/latest/vcs.html#accessing-repositories-from-hosted-weblate

As I understand it, if you are using hosted weblate, there is already a weblate github user for hosted weblate, and I think you need to give that user access to your repo as a committer.

If you are hosting your own, as we are on weblate.osgeo.org, then you use the ssh key we have - listed here -  https://weblate.osgeo.org/keys/

and you add that key as a deploy key in your project settings, and make sure it has permission to push.  I got it wrong a couple of times.  Anyway will see if I can add to the workshop one.

**Comment by robe2** [comment]
> Well, my experience with Weblate is that I couldn't manage to get it working to push to Github. But one reason was probably, that I didn't have enough permissions on the Github repository. So I will try now again with the repository I have admin access for.
> 
> The hosted Weblate for Libre Projects has some more version systems listed, for example Github: ![Screenshot from 2021-11-29 15-44-01](https://user-images.githubusercontent.com/227762/143820783-2e821282-b924-42f6-bb82-eff62272f396.png) Not sure if this is a configuration option. Eventually it simplifies the setup with Github.

Hmm where are you seeing that? 
If it's a new feature.  I think hosted weblate was running a newer version.  I've since upgraded ours to 4.9.1 which should be latest stable.  Though I think 4.10 might have come out or is about to come out.

Also might have to do with the allowed authentications.  As mentioned I do plan to open it up to gitlab and github users once I've tested and no screams from @strk 

@strk raised a concern, depends on our workflow how much we open it up.  So will probably test for a bit with just OSGeo users.

**Comment by robe2** [comment]
@dkastl   Okay I got the same error, I guess I got to define an OSGeo Weblate account separate from my own as it allows deploy only once.

**Comment by robe2** [comment]
@dkastl There is now an @osgeoweblate   can you try inviting that to contribute.  I installed the SSH Key there but not sure if I need to turn on 2-factor.

**Comment by robe2** [comment]
@dkastl  I got it to work.  strk did an edit and it came thru.  Though I manually clicked the push cause I was impatient.

https://github.com/postgis/postgis-workshops/commit/f3b1f810a9c25d2889c2216b62d4843a44544858

**Comment by dkastl** [comment]
@robe2 , thanks for setting up the Weblate user on Github! This seems to be the only way, because SSH keys can be only used once in Github.

My problem now is, that I require 2FA for all the organisations. So if the OSGeo Weblate account does not use 2FA it will not work. I sent an invitation though to try.

**Comment by robe2** [comment]
@dkastl I didn't see any notifications in the @osgeoweblate account.  I just enabled 2-factor on the account.  Can you try again?

**Comment by dkastl** [comment]
Thanks, @robe2 !
The invitation was still pending, so I cancelled it and created a new invitation. Did you receive a notification this time?

**Comment by osgeoweblate** [comment]
@dkastl  no didn't see it.  It's possible I was over zealous in turning off notifications.

Which project is it in?  I thought maybe this one  - https://github.com/gtt-project/redmine_gtt/invitations but says no invitations when I'm logged into osgeoweblate account.  I did get the notification for postgis-workshop but that was before I turned off most notifications.

**Comment by osgeoweblate** [comment]
@dkastl  disregard just found it in my mail and accepted.

**Comment by dkastl** [comment]
Oh, great! Now I can find the user in the team I created for that under the organisation level, because I wanted to avoid that this invitation is necessary for every repository. 

Thanks for the help! It would be nice this could be automated similar to how hosted Weblate does, but I don't know yet how. I also try to find out, if this user can be granted only the necessary permissions to for example submit pull requests instead of directly updating the development branch.

**Comment by robe2** [comment]
I'll give it a stab later today.  Looks like I need to add some extra configurations to the weblate config to make that work.

https://docs.weblate.org/en/latest/vcs.html#github-pull-requests

**Comment by robe2** [comment]
@dkastl  can you see if you can do it now.  I put in the config for osgeoweblate account and setup a development key.  I do see Github as an option in VCS control, but forget if I so it before I put those in.

**Comment by dkastl** [comment]
Thanks, @robe2 !
I can confirm that `Github` was not available as an option before and it is now there.

Will give it a try with another plugin repository the next days. So far Webhooks and automatic merges worked well, but as also discussed on the PostGIS ML, it's better to merge translations as pull requests to not accidentally break stuff.

**Comment by robe2** [comment]
@dkastl Nice I see you were able to enable github pull requests.

https://github.com/gtt-project/redmine_decision_tree/pull/3 

Is it a simple matter of just taking out the push url? and using Github for the VCS instead of git? or other things need to be done?

**Comment by robe2** [comment]
Oh feel free to update these instructions I started putting together - 
https://wiki.osgeo.org/wiki/SAC:Weblate

**Comment by dkastl** [comment]
Yes, it's as simple as that ... it seems. 

And I think I could also remove the `@osgeoweblate` account from the organisation now. I added the user to a team, that had granted write access to the first repositories. But I already removed the access to the repositories and it didn't affect the ability to create pull requests ... as expected. \
This would make everything a lot simpler, no need to care about 2FA and no need to grant a user write access at all.
